### ISSUE TITLE
Add ability to add custom TokenReplayValidator or TokenReplayCache

### ISSUE THREAD

**Issue by tanukus** [ITfoxtec/ITfoxtec.Identity.Saml2#72]
@Revsgaard

I reviewed the code in `Saml2ResponseSecurityTokenHandler.cs` although there is a `DetectReplayedTokens` config setting that gets into the TokenValidationParametes, the default behavior for token replay validation is to bypass the validation if there are no validators and no token replay cache is set.

However there is no way the code lets you set those values, they are not exposed anywhere.

```
if (TokenValidationParameters.ValidateTokenReplay)
{      
    ValidateTokenReplay(saml2SecurityToken.Assertion.Conditions.NotBefore, tokenString, TokenValidationParameters); 
}
```

There is no way I could set the Replay cache or validator in the TokenValidationParameters as it is not exposed anywhere. So, the `DetectReplayedTokens` setting is of not much help.

This method in `SAML2Request.cs` needs an update? 

`public static Saml2IdentityConfiguration GetIdentityConfiguration(Saml2Configuration config)`

Thanks for your work on the library and any insight into this issue.

**Comment by Revsgaard** [comment]
Thank you for your contribution, I have merge your pull request. 

I do not think you need the IssuerTokenResolver. I think you need the following instead:

**.NET Framework** 

    public IdentityModelCaches TokenReplayCache { get; set; }
    public TimeSpan? TokenReplayCacheExpirationPeriod { get; set; }

https://github.com/ITfoxtec/ITfoxtec.Identity.Saml2/blob/master/src/ITfoxtec.Identity.Saml2/Configuration/Saml2Configuration.cs#L44

**.NET Core / .NET 5.0** 

    public ITokenReplayCache TokenReplayCache { get; set; }

https://github.com/ITfoxtec/ITfoxtec.Identity.Saml2/blob/master/src/ITfoxtec.Identity.Saml2/Configuration/Saml2Configuration.cs#L47

Please look at the code and see if you agree.

**Comment by tanukus** [comment]
@Revsgaard Looks good, that exactly what I did in my code. I'll rollback my code and update with your changes. Are you planning to push the NuGet package soon?

**Comment by tanukus** [comment]
Also, I have a default implementation of the ITokenReplayCache for .NET Core / .NET 5. You could use it if you like it, thanks.

```
    /// <summary>
    /// Distributed implementation for ITokenReplayCache with cache expiration of 30 minutes.
    /// The SAML 2 token's expiration i.e. "SessionNotOnOrAfter" should be less than 30 minutes 
    /// to prevent a replay attack, when using this default cache.
    /// </summary>
    internal sealed class Saml2TokenReplayDistributedCache : ITokenReplayCache
    {
        private static byte[] _dummyData = Encoding.UTF8.GetBytes("dummy");
        private readonly IDistributedCache _cache;

        public Saml2TokenReplayDistributedCache(IDistributedCache cache)
        {
            _cache = cache;
        }
        
        public bool TryFind(string securityToken)
        {
            CheckToken(securityToken);
            var token = ReadAssertionToken(securityToken);
            var result = _cache.Get(token.Id);
            return result != null;
        }

        public bool TryAdd(string securityToken, DateTime expiresOn)
        {
            CheckToken(securityToken);
            var token = ReadAssertionToken(securityToken);
            _cache.Set(token.Id, _dummyData, new DistributedCacheEntryOptions
            {
                AbsoluteExpirationRelativeToNow = TimeSpan.FromMinutes(30)
            });
            return true;
        }

        private void CheckToken(string securityToken)
        {
            if (string.IsNullOrWhiteSpace(securityToken))
                throw new Saml2Exception($"Invalid security token:{securityToken}");
        }

        private Saml2SecurityToken ReadAssertionToken(string securityToken)
        {
            return new Saml2ResponseSecurityTokenHandler().ReadSaml2Token(securityToken);
        }
    }
```

**Comment by Revsgaard** [comment]
Thank you for your reply and the ITokenReplayCache implementation. Maybe I can use it at some point or someone else.

I'll release the change sune either in an ordinary version or beta version.

**Comment by Revsgaard** [comment]
The feature is released in version [4.4.0-beta1]([url](https://github.com/ITfoxtec/ITfoxtec.Identity.Saml2/releases/tag/4.4.0-beta1)).
